### DISCUSSION TITLE
Can uv_fs_t be relocated?

### DISCUSSION THREAD

**Discussion by nlw0** [libuv/libuv#3595] [Ideas]
I'm doing some experiments calling libuv filesystem functions from a garbage-collected language. Everything works fine if I `malloc` the `uv_fs_t` structure and hold a static pointer. It seems to break if rely on GC-managed memory, and I'm assuming it's because we cannot relocate the structures.

I guess it's not surprising, but I was curious to ask: are there reasons it has to be this way, or is there a chance the library could be modified to support relocatable structures one day? Is this something that will probably never change, or might it just depend on some refactoring?

**Comment by vtjnash** [comment]
How is libuv supposed to know it moved if it is in use at the time on a worker thread?

**Reply by nlw0** [reply]
The idea is libuv should _not_ know it moved. We would move around the `uv_fs_t` "req" struct, and everything would just work. When eg. Julia calls libuv, it keeps the memory in place, and libuv shouldn't know if it moved in the meantime. If no pointers to this struct are kept within libuv, this should work. In Julia, today we allocate memory with `malloc` for this structure, then `free` it later. I'm wondering if we could use a Julia GC-managed `Vector` instead. It didn't work when I tried it, so I'm trying to figure out if it was even possible in the first place.

**Comment by nlw0** [comment]
I must have done something wrong, it actually does seem to work! Sorry for the noise.

**Reply by bnoordhuis** [reply]
`uv _fs_t` and other `uv_req_t` instances are linked into one or more intrusive linked lists. Moving them isn't safe.

**Reply by nlw0** [reply]
OK, thanks @bnoordhuis
